runtime/src/main/java/org/corfudb/runtime/collections/Query.java

```diff
@@ -0,0 +1,424 @@
+ *
+ * Created by zlokhandwala on 2019-08-09.
+ */
+public class Query {
+
+    private final TableRegistry tableRegistry;
+
+    private final ObjectsView objectsView;
+
+    private final String namespace;
+
+    /**
+     * Creates a Query interface.
+     *
+     * @param tableRegistry Table registry from the corfu client.
+     * @param objectsView   Objects View from the corfu client.
+     * @param namespace     Namespace to perform the queries within.
+     */
+    public Query(final TableRegistry tableRegistry, final ObjectsView objectsView, final String namespace) {
+        this.tableRegistry = tableRegistry;
+        this.objectsView = objectsView;
+        this.namespace = namespace;
+    }
+
+    private Token getToken(Timestamp timestamp) {
+        return new Token(timestamp.getEpoch(), timestamp.getSequence());
+    }
+
+    private void txBegin(Timestamp timestamp) {
+        TransactionBuilder transactionBuilder = objectsView.TXBuild()
+                .type(TransactionType.SNAPSHOT);
+
+        if (timestamp != null) {
+            transactionBuilder.snapshot(getToken(timestamp));
+        }
+
+        transactionBuilder.build().begin();
+    }
+
+    private void txEnd() {
+        objectsView.TXEnd();
+    }
+
+    private <K extends Message, V extends Message, M extends Message>
+    Table<K, V, M> getTable(@Nonnull final String tableName) {
+        return tableRegistry.getTable(this.namespace, tableName);
+    }
+
+    /**
+     * Fetch the CorfuRecord for the specified key at the LATEST timestamp (default).
+     *
+     * @param tableName Table name.
+     * @param key       Key.
+     * @param <K>       Type of key.
+     * @param <V>       Type of the payload/value.
+     * @param <M>       Type of the metadata.
+     * @return CorfuRecord encapsulating the payload and the metadata.
+     */
+    @Nullable
+    public <K extends Message, V extends Message, M extends Message>
+    CorfuRecord<V, M> getRecord(@Nonnull final String tableName,
+                                @Nonnull final K key) {
+        return getRecord(tableName, null, key);
+    }
+
+    /**
+     * Fetch the Value payload for the specified key at the specified snapshot/timestamp.
+     *
+     * @param tableName Table name.
+     * @param timestamp Timestamp to perform the query.
+     * @param key       Key.
+     * @param <K>       Type of key.
+     * @param <V>       Type of value/payload.
+     * @param <M>       Type of metadata.
+     * @return Value.
+     */
+    @Nullable
+    public <K extends Message, V extends Message, M extends Message>
+    CorfuRecord<V, M> getRecord(@Nonnull final String tableName,
+                          @Nullable final Timestamp timestamp,
+                          @Nonnull final K key) {
+        try {
+            txBegin(timestamp);
+            return ((Table<K, V, M>) getTable(tableName)).get(key);
```

Comment:
Looks like NPE cold happen here

Reply:
Good catch! thanks Slava!

Reply:
In fact let me add a test case for this !